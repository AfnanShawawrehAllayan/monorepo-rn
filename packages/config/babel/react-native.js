module.exports = {
  presets: ['@react-native/babel-preset'],
  plugins: [
    [
      'module-resolver',
      {
        root: ['.'],
        extensions: [
          '.ios.ts',
          '.android.ts',
          '.ts',
          '.ios.tsx',
          '.android.tsx',
          '.tsx',
          '.jsx',
          '.js',
          '.json',
        ],
        alias: {
          '@components': './src/components',
          '@screens': './src/screens',
          '@navigation': './src/navigation',
          '@utils': './src/utils',
          '@hooks': './src/hooks',
          '@assets': './src/assets',
          '@config': './src/config',
        },
      },
    ],
    'react-native-reanimated/plugin',
  ],
  env: {
    production: {
      plugins: ['transform-remove-console'],
    },
  },
};
