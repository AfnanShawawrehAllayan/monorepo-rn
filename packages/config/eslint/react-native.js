module.exports = {
  root: true,
  extends: [
    'eslint:recommended',
    'plugin:@typescript-eslint/recommended',
    'plugin:react/recommended',
    'plugin:react-hooks/recommended',
    'plugin:react-native/all',
    'plugin:import/errors',
    'plugin:import/warnings',
    'plugin:import/typescript',
    'plugin:jest/recommended',
    'prettier',
  ],
  parser: '@typescript-eslint/parser',
  parserOptions: {
    ecmaFeatures: {
      jsx: true,
    },
    ecmaVersion: 2021,
    sourceType: 'module',
  },
  plugins: ['@typescript-eslint', 'react', 'react-native', 'import', 'jest'],
  settings: {
    react: {
      version: 'detect',
    },
    'import/resolver': {
      typescript: {},
      node: {
        extensions: ['.js', '.jsx', '.ts', '.tsx'],
      },
    },
  },
  rules: {
    // React Native specific rules
    'react-native/no-unused-styles': 'error',
    'react-native/split-platform-components': 'error',
    'react-native/no-inline-styles': 'warn',
    'react-native/no-color-literals': 'warn',
    'react-native/no-raw-text': ['warn', { skip: ['Text'] }],
    'react-native/no-single-element-style-arrays': 'error',

    // TypeScript rules
    '@typescript-eslint/explicit-module-boundary-types': 'warn',
    '@typescript-eslint/no-explicit-any': 'warn',
    '@typescript-eslint/no-unused-vars': ['error', { argsIgnorePattern: '^_' }],
    '@typescript-eslint/consistent-type-imports': ['error', { prefer: 'type-imports' }],

    // React rules
    'react/prop-types': 'off',
    'react/react-in-jsx-scope': 'off',
    'react/display-name': 'off',
    'react/jsx-curly-brace-presence': ['error', { props: 'never', children: 'never' }],

    // Import rules
    'import/order': [
      'error',
      {
        groups: ['builtin', 'external', 'internal', 'parent', 'sibling', 'index'],
        'newlines-between': 'always',
        alphabetize: {
          order: 'asc',
          caseInsensitive: true,
        },
      },
    ],
    'import/no-duplicates': 'error',
    'import/no-unresolved': 'error',
    'import/no-cycle': 'error',

    // Jest rules
    'jest/no-disabled-tests': 'warn',
    'jest/no-focused-tests': 'error',
    'jest/no-identical-title': 'error',
    'jest/valid-expect': 'error',
  },
  env: {
    'react-native/react-native': true,
    jest: true,
    'jest/globals': true,
  },
  overrides: [
    {
      files: ['*.test.ts', '*.test.tsx', '*.spec.ts', '*.spec.tsx'],
      rules: {
        '@typescript-eslint/no-explicit-any': 'off',
        'react-native/no-raw-text': 'off',
      },
    },
    {
      files: ['*.js', '*.jsx'],
      rules: {
        '@typescript-eslint/explicit-module-boundary-types': 'off',
      },
    },
  ],
};
