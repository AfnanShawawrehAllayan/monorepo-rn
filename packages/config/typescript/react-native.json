{
  "$schema": "https://json.schemastore.org/tsconfig",
  "display": "React Native",
  "compilerOptions": {
    "target": "esnext",
    "module": "esnext",
    "lib": ["esnext"],
    "allowJs": true,
    "jsx": "react-native",
    "noEmit": true,
    "isolatedModules": true,
    "strict": true,
    "moduleResolution": "node",
    "allowSyntheticDefaultImports": true,
    "esModuleInterop": true,
    "skipLibCheck": true,
    "resolveJsonModule": true,
    "baseUrl": ".",
    "paths": {
      "*": ["*", "*.ios", "*.android"]
    },
    "forceConsistentCasingInFileNames": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "noImplicitReturns": true,
    "noFallthroughCasesInSwitch": true,
    "noUncheckedIndexedAccess": true,
    "noImplicitOverride": true,
    "noPropertyAccessFromIndexSignature": true,
    "exactOptionalPropertyTypes": true,
    "types": ["jest", "react-native"]
  },
  "exclude": ["node_modules", "babel.config.js", "metro.config.js", "jest.config.js"]
}
