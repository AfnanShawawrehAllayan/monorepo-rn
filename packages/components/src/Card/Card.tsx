import React from 'react';
import { View, StyleSheet, ViewStyle } from 'react-native';
import { useScale } from 'hooks';
import { useTheme } from 'theme';

interface CardProps {
  children: React.ReactNode;
  style?: ViewStyle;
  elevation?: number;
  outlined?: boolean;
  padded?: boolean;
}

export const Card = ({
  children,
  style,
  elevation = 2,
  outlined = false,
  padded = true,
}: CardProps): React.JSX.Element => {
  const theme = useTheme();
  const { scale } = useScale();

  const styles = StyleSheet.create({
    card: {
      borderRadius: scale(8),
      overflow: 'hidden',
    },
  });

  const cardStyles = [
    styles.card,
    {
      backgroundColor: theme.colors.card,
      padding: padded ? scale(16) : 0,
      borderColor: outlined ? theme.colors.border : 'transparent',
      borderWidth: outlined ? 1 : 0,
      ...getShadow(elevation, theme.mode === 'dark'),
    },
    style,
  ];

  return <View style={cardStyles}>{children}</View>;
};

// Shadow utility function that works on iOS and Android
const getShadow = (elevation: number, isDark: boolean): ViewStyle => {
  const shadowColor = isDark ? '#000' : '#000';
  const shadowOpacity = isDark ? 0.3 : 0.1;

  return {
    shadowColor,
    shadowOffset: {
      width: 0,
      height: elevation,
    },
    shadowOpacity,
    shadowRadius: elevation * 0.8,
    elevation,
  };
}; 