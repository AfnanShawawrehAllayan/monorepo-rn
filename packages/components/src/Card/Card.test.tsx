// Mock modules before imports
jest.mock('react-native/Libraries/Utilities/Dimensions', () => ({
  get: jest.fn().mockReturnValue({ width: 375, height: 812 }),
}));

jest.mock('hooks', () => ({
  useScale: () => ({
    scale: jest.fn((value) => value),
    verticalScale: jest.fn((value) => value),
    moderateScale: jest.fn((value) => value),
  }),
}));

// Mock theme
jest.mock('theme', () => ({
  useTheme: () => ({
    mode: 'light',
    colors: {
      card: '#FFFFFF',
      border: '#E0E0E0',
      primary: '#007AFF',
      secondary: '#5856D6',
      background: '#FFFFFF',
      surface: '#F5F5F5',
      text: '#000000',
    },
    spacing: {
      xs: 4,
      s: 8,
      m: 16,
      l: 24,
      xl: 32,
    },
  }),
}));

import React from 'react';
import { render } from '@testing-library/react-native';
import { Text, View } from 'react-native';
import { Card } from './Card';

describe('Card Component', () => {
  it('renders correctly with default props', () => {
    const { toJSON } = render(
      <Card>
        <Text>Card Content</Text>
      </Card>
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders with custom style', () => {
    const customStyle = { marginTop: 20 };
    const { toJSON } = render(
      <Card style={customStyle}>
        <Text>Card Content</Text>
      </Card>
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders with custom elevation', () => {
    const { toJSON } = render(
      <Card elevation={5}>
        <Text>Card Content</Text>
      </Card>
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders with outline', () => {
    const { toJSON } = render(
      <Card outlined>
        <Text>Card Content</Text>
      </Card>
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders without padding', () => {
    const { toJSON } = render(
      <Card padded={false}>
        <Text>Card Content</Text>
      </Card>
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders with complex content', () => {
    const { toJSON } = render(
      <Card>
        <View>
          <Text>Title</Text>
          <Text>Description</Text>
          <View>
            <Text>Nested Content</Text>
          </View>
        </View>
      </Card>
    );
    expect(toJSON()).toMatchSnapshot();
  });
}); 