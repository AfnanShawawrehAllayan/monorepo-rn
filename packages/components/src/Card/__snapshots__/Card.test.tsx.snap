// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Component renders correctly with default props 1`] = `
<View
  style={
    [
      {
        "borderRadius": 8,
        "overflow": "hidden",
      },
      {
        "backgroundColor": "#FFFFFF",
        "borderColor": "transparent",
        "borderWidth": 0,
        "elevation": 2,
        "padding": 16,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.1,
        "shadowRadius": 1.6,
      },
      undefined,
    ]
  }
>
  <Text>
    Card Content
  </Text>
</View>
`;

exports[`Card Component renders with complex content 1`] = `
<View
  style={
    [
      {
        "borderRadius": 8,
        "overflow": "hidden",
      },
      {
        "backgroundColor": "#FFFFFF",
        "borderColor": "transparent",
        "borderWidth": 0,
        "elevation": 2,
        "padding": 16,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.1,
        "shadowRadius": 1.6,
      },
      undefined,
    ]
  }
>
  <View>
    <Text>
      Title
    </Text>
    <Text>
      Description
    </Text>
    <View>
      <Text>
        Nested Content
      </Text>
    </View>
  </View>
</View>
`;

exports[`Card Component renders with custom elevation 1`] = `
<View
  style={
    [
      {
        "borderRadius": 8,
        "overflow": "hidden",
      },
      {
        "backgroundColor": "#FFFFFF",
        "borderColor": "transparent",
        "borderWidth": 0,
        "elevation": 5,
        "padding": 16,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 5,
          "width": 0,
        },
        "shadowOpacity": 0.1,
        "shadowRadius": 4,
      },
      undefined,
    ]
  }
>
  <Text>
    Card Content
  </Text>
</View>
`;

exports[`Card Component renders with custom style 1`] = `
<View
  style={
    [
      {
        "borderRadius": 8,
        "overflow": "hidden",
      },
      {
        "backgroundColor": "#FFFFFF",
        "borderColor": "transparent",
        "borderWidth": 0,
        "elevation": 2,
        "padding": 16,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.1,
        "shadowRadius": 1.6,
      },
      {
        "marginTop": 20,
      },
    ]
  }
>
  <Text>
    Card Content
  </Text>
</View>
`;

exports[`Card Component renders with outline 1`] = `
<View
  style={
    [
      {
        "borderRadius": 8,
        "overflow": "hidden",
      },
      {
        "backgroundColor": "#FFFFFF",
        "borderColor": "#E0E0E0",
        "borderWidth": 1,
        "elevation": 2,
        "padding": 16,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.1,
        "shadowRadius": 1.6,
      },
      undefined,
    ]
  }
>
  <Text>
    Card Content
  </Text>
</View>
`;

exports[`Card Component renders without padding 1`] = `
<View
  style={
    [
      {
        "borderRadius": 8,
        "overflow": "hidden",
      },
      {
        "backgroundColor": "#FFFFFF",
        "borderColor": "transparent",
        "borderWidth": 0,
        "elevation": 2,
        "padding": 0,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.1,
        "shadowRadius": 1.6,
      },
      undefined,
    ]
  }
>
  <Text>
    Card Content
  </Text>
</View>
`;
